src/main/host/descriptor/mod.rs

```diff
@@ -44,6 +58,53 @@ bitflags::bitflags! {
     }
 }
 
+impl FileMode {
+    /// Returns an empty `Err` if the file mode is not valid (neither read nor write).
+    pub fn as_o_flags(&self) -> Result<OFlag, ()> {
+        if self.contains(Self::READ | Self::WRITE) {
+            return Ok(OFlag::O_RDWR);
+        }
+        if self.contains(Self::READ) {
+            return Ok(OFlag::O_RDONLY);
+        }
+        if self.contains(Self::WRITE) {
+            return Ok(OFlag::O_WRONLY);
+        }
+        // O_RDONLY is 0, so we should indicate this state with an error rather
+        // than returning 0
+        return Err(());
+    }
+
+    /// Returns a tuple of the `FileMode` and any remaining flags, or an empty `Err` if
+    /// the flags aren't valid (for example specifying both `O_RDWR` and `O_RDONLY`).
+    pub fn from_o_flags(flags: OFlag) -> Result<(Self, OFlag), ()> {
+        // O_RDONLY is 0, so some of the following checks are redundant
+        let has_rdwr = flags.contains(OFlag::O_RDWR);
+        let has_rdonly = flags.contains(OFlag::O_RDONLY);
+        let has_wronly = flags.contains(OFlag::O_WRONLY);
+
+        if has_rdwr {
```

Comment:
Alternatively:

```
match (has_rdwr, has_rdonly, has_wronly) {
  (true, false, false) => Ok(...),
  (false, true, false) => Ok(...),
  (false, fasle, true) => Ok(...),
  _ => Err(())
}
```

Reply:
I like this format, but I found an error in my existing code. Since `O_RDONLY` is 0, `flags.contains(OFlag::O_RDONLY)` is always true, which is not what we want. So I fixed the logic in this function, and also added unit tests to make sure it's handled properly. For example `O_RDONLY | O_WRONLY == FileMode::WRITE` since `O_RDONLY` is 0.

Reply:
Whoops, good catch!